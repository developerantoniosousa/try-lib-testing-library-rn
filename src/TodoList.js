import React from 'react';
import { View, TextInput, Text, Button } from 'react-native';

const TodoList = ({ defaultTodos }) => {
  const [todos, setTodos] = React.useState(defaultTodos);
  const [todoInputvalue, setTodoInputValue] = React.useState('');

  function handleAddTodo() {
    if (!todoInputvalue) {
      return;
    }

    setTodos([...todos, todoInputvalue]);
    setTodoInputValue('');
  }

  return (
    <View>
      <TextInput
        testID="input-add-todo"
        value={todoInputvalue}
        onChangeText={setTodoInputValue}
      />
      <Button testID="button-add-todo" title="Add" onPress={handleAddTodo} />
      <View testID="todo-list">
        {todos.map((todo) => (
          <Text testID="todo-item" key={todo}>
            {todo}
          </Text>
        ))}
      </View>
    </View>
  );
};

TodoList.defaultProps = {
  defaultTodos: [],
};

export default TodoList;
