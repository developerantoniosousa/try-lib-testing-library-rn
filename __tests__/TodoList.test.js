import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import TodoList from '../src/TodoList';

const todos = ['do my homework', 'make the bed'];

describe('TodoList', () => {
  it('should render corretly without pass any prop to element', () => {
    const tree = render(<TodoList />);
    expect(tree).toMatchSnapshot();
  });

  it('should render component with the defaultTodos passed as a prop to element correctly', () => {
    const tree = render(<TodoList defaultTodos={todos} />);
    expect(tree).toMatchSnapshot();
  });

  it('should block the insertion of todo because the todoInputvalue is empty', () => {
    const { getByTestId } = render(<TodoList />);
    fireEvent.press(getByTestId('button-add-todo'));
    expect(getByTestId('todo-list').children).toHaveLength(0);
  });

  it('should render one todo in the list', () => {
    const { getByTestId, getAllByTestId } = render(<TodoList />)
    fireEvent.changeText(getByTestId('input-add-todo'), todos[0]);
    fireEvent.press(getByTestId('button-add-todo'));
    fireEvent.changeText(getByTestId('input-add-todo'), todos[1]);
    fireEvent.press(getByTestId('button-add-todo'));
    expect(getByTestId('todo-list').children).toHaveLength(2);
    const todoItems = getAllByTestId('todo-item')
    expect(todoItems[0].children[0]).toBe(todos[0])
    expect(todoItems[1].children[0]).toBe(todos[1])
  });

  it('should clear the input when add a new todo', () => {
    const { getByTestId } = render(<TodoList />);
    const textInput = getByTestId('input-add-todo');
    fireEvent.changeText(textInput, todos[0]);
    expect(textInput).toHaveProp('value', todos[0]);
    fireEvent.press(getByTestId('button-add-todo'));
    expect(textInput).toHaveProp('value', '');
  });

  it('should render a list of todos passed by defaultTodos prop and in correct order', () => {
    const { getByTestId } = render(<TodoList defaultTodos={todos} />);
    expect(getByTestId('todo-list').children).toHaveLength(2);
    expect(getByTestId('todo-list').children[0]).toHaveProp('children', todos[0]);
    expect(getByTestId('todo-list').children[1]).toHaveProp('children', todos[1]);
  });
});
