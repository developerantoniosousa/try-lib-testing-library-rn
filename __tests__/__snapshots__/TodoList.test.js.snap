// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render component with the defaultTodos passed as a prop to element correctly 1`] = `
Object {
  "asJSON": [Function],
  "baseElement": <View
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={true}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <TextInput
          allowFontScaling={true}
          rejectResponderTermination={true}
          testID="input-add-todo"
          underlineColorAndroid="transparent"
          value=""
        />
        <Button
          testID="button-add-todo"
          title="Add"
        />
        <View
          testID="todo-list"
        >
          <Text
            testID="todo-item"
          >
            do my homework
          </Text>
          <Text
            testID="todo-item"
          >
            make the bed
          </Text>
        </View>
      </View>
    </View>
  </View>,
  "container": <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <TextInput
        allowFontScaling={true}
        rejectResponderTermination={true}
        testID="input-add-todo"
        underlineColorAndroid="transparent"
        value=""
      />
      <Button
        testID="button-add-todo"
        title="Add"
      />
      <View
        testID="todo-list"
      >
        <Text
          testID="todo-item"
        >
          do my homework
        </Text>
        <Text
          testID="todo-item"
        >
          make the bed
        </Text>
      </View>
    </View>
  </View>,
  "debug": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByHintText": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByDisplayValue": [Function],
  "findByHintText": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByHintText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByDisplayValue": [Function],
  "getByHintText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByHintText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByDisplayValue": [Function],
  "queryByHintText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TodoList should render corretly without pass any prop to element 1`] = `
Object {
  "asJSON": [Function],
  "baseElement": <View
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={true}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <TextInput
          allowFontScaling={true}
          rejectResponderTermination={true}
          testID="input-add-todo"
          underlineColorAndroid="transparent"
          value=""
        />
        <Button
          testID="button-add-todo"
          title="Add"
        />
        <View
          testID="todo-list"
        />
      </View>
    </View>
  </View>,
  "container": <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <TextInput
        allowFontScaling={true}
        rejectResponderTermination={true}
        testID="input-add-todo"
        underlineColorAndroid="transparent"
        value=""
      />
      <Button
        testID="button-add-todo"
        title="Add"
      />
      <View
        testID="todo-list"
      />
    </View>
  </View>,
  "debug": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByHintText": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByDisplayValue": [Function],
  "findByHintText": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByHintText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByDisplayValue": [Function],
  "getByHintText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByHintText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByDisplayValue": [Function],
  "queryByHintText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
